# Regression Week 5: Feature Selection and LASSO (Interpretation)

In this notebook, you will use LASSO to select features, building on a pre-implemented solver for LASSO (using Turi Create, though you can use other solvers). You will:
* Run LASSO with different L1 penalties.
* Choose best L1 penalty using a validation set.
* Choose best L1 penalty using a validation set, with additional constraint on the size of subset.

In the second notebook, you will implement your own LASSO solver, using coordinate descent. 

# Fire up Turi Create

In [1]:
import turicreate

The history saving thread hit an unexpected error (DatabaseError('database disk image is malformed',)).History will not be written to the database.


# Load in house sales data

Dataset is from house sales in King County, the region where the city of Seattle, WA is located.

In [2]:
sales = turicreate.SFrame('home_data.sframe/')

# Create new features

As in Week 2, we consider features that are some transformations of inputs.

In [3]:
from math import log, sqrt
sales['sqft_living_sqrt'] = sales['sqft_living'].apply(sqrt)
sales['sqft_lot_sqrt'] = sales['sqft_lot'].apply(sqrt)
sales['bedrooms_square'] = sales['bedrooms']*sales['bedrooms']

# In the dataset, 'floors' was defined with type string, 
# so we'll convert them to float, before creating a new feature.
sales['floors'] = sales['floors'].astype(float) 
sales['floors_square'] = sales['floors']*sales['floors']

* Squaring bedrooms will increase the separation between not many bedrooms (e.g. 1) and lots of bedrooms (e.g. 4) since 1^2 = 1 but 4^2 = 16. Consequently this variable will mostly affect houses with many bedrooms.
* On the other hand, taking square root of sqft_living will decrease the separation between big house and small house. The owner may not be exactly twice as happy for getting a house that is twice as big.

# Learn regression weights with L1 penalty

Let us fit a model with all the features available, plus the features we just created above.

In [4]:
all_features = ['bedrooms', 'bedrooms_square',
                'bathrooms',
                'sqft_living', 'sqft_living_sqrt',
                'sqft_lot', 'sqft_lot_sqrt',
                'floors', 'floors_square',
                'waterfront', 'view', 'condition', 'grade',
                'sqft_above',
                'sqft_basement',
                'yr_built', 'yr_renovated']

Applying L1 penalty requires adding an extra parameter (`l1_penalty`) to the linear regression call in Turi Create. (Other tools may have separate implementations of LASSO.)  Note that it's important to set `l2_penalty=0` to ensure we don't introduce an additional L2 penalty.

In [5]:
model_all = turicreate.linear_regression.create(sales, target='price', features=all_features,
                                                validation_set=None, 
                                                l2_penalty=0., l1_penalty=1e10)

Linear regression:

--------------------------------------------------------

Number of examples          : 21613

Number of features          : 17

Number of unpacked features : 17

Number of coefficients    : 18

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+--------------------+---------------------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Max Error | Training Root-Mean-Square Error |

+-----------+----------+-----------+--------------+--------------------+---------------------------------+

| 0         | 1        | 1.000000  | 1.028215     | 7700000.000000     | 653047.733994                   |

| 1         | 2        | 0.000002  | 1.904030     | 6962915.603493     | 426631.749026                   |

| 2         | 3        | 0.000002  | 1.939326     | 6843144.200219     | 392488.929838                   |

| 3         | 4        | 0.000002  | 1.989875     | 6831900.032123     | 385340.166783                   |

| 4         | 5        | 0.000002  | 2.039933     | 6847166.848958     | 384842.383767                   |

| 5         | 6        | 0.000002  | 2.119823     | 6869667.895833     | 385998.458623                   |

| 10        | 11       | 0.000002  | 2.542021     | 6842123.232651     | 364204.576180                   |

+-----------+----------+-----------+--------------+--------------------+---------------------------------+

Completed (Iteration limit reached).

This model may not be optimal. To improve it, consider increasing `max_iterations`.

Find what features had non-zero weight.

In [6]:
rows = len(all_features)
model_all.coefficients.print_rows(num_rows=rows)

+------------------+-------+--------------------+--------+
|       name       | index |       value        | stderr |
+------------------+-------+--------------------+--------+
|   (intercept)    |  None | 274873.0559504957  |  None  |
|     bedrooms     |  None |        0.0         |  None  |
| bedrooms_square  |  None |        0.0         |  None  |
|    bathrooms     |  None | 8468.531086910072  |  None  |
|   sqft_living    |  None | 24.420720982445214 |  None  |
| sqft_living_sqrt |  None | 350.0605533860648  |  None  |
|     sqft_lot     |  None |        0.0         |  None  |
|  sqft_lot_sqrt   |  None |        0.0         |  None  |
|      floors      |  None |        0.0         |  None  |
|  floors_square   |  None |        0.0         |  None  |
|    waterfront    |  None |        0.0         |  None  |
|       view       |  None |        0.0         |  None  |
|    condition     |  None |        0.0         |  None  |
|      grade       |  None | 842.0680348976282  |  None 

Note that a majority of the weights have been set to zero. So by setting an L1 penalty that's large enough, we are performing a subset selection. 

***QUIZ QUESTION***:
According to this list of weights, which of the features have been chosen? 

# Selecting an L1 penalty

To find a good L1 penalty, we will explore multiple values using a validation set. Let us do three way split into train, validation, and test sets:
* Split our sales data into 2 sets: training and test
* Further split our training data into two sets: train, validation

Be *very* careful that you use seed = 1 to ensure you get the same answer!

In [7]:
(training_and_validation, testing) = sales.random_split(.9,seed=1) # initial train/test split
(training, validation) = training_and_validation.random_split(0.5, seed=1) # split training into train and validate

Next, we write a loop that does the following:
* For `l1_penalty` in [10^1, 10^1.5, 10^2, 10^2.5, ..., 10^7] (to get this in Python, type `np.logspace(1, 7, num=13)`.)
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list.
    * Compute the RSS on VALIDATION data (here you will want to use `.predict()`) for that `l1_penalty`
* Report which `l1_penalty` produced the lowest RSS on validation data.

When you call `linear_regression.create()` make sure you set `validation_set = None`.

Note: you can turn off the print out of `linear_regression.create()` with `verbose = False`

In [8]:
import numpy as np

np_logspace = np.logspace(1, 7, num = 13)

num_penality = len(np_logspace)

RSS = np.empty(num_penality)

for i in range(num_penality):
    
    model = turicreate.linear_regression.create(training, target = 'price', 
                                                features = all_features, 
                                                l1_penalty = np_logspace[i],
                                                l2_penalty = 0,
                                                validation_set = None,
                                                verbose = False)
    
    prediction = model.predict(validation)
    
    errors = prediction - validation['price']
    
    RSS[i] = np.dot(errors,errors)
    
print (RSS)

[6.25766285e+14 6.25766285e+14 6.25766286e+14 6.25766288e+14
 6.25766295e+14 6.25766317e+14 6.25766387e+14 6.25766607e+14
 6.25767303e+14 6.25769508e+14 6.25776518e+14 6.25799063e+14
 6.25883719e+14]


In [9]:
rss_min = RSS.min()

for i in range(num_penality):
    
    if RSS[i]==rss_min:
        
        print ("Penalty: " + str(np_logspace[i]) + "\nRSS: " + str(RSS[i]))

Penalty: 10.0
RSS: 625766285142460.6


*** QUIZ QUESTION. *** What was the best value for the `l1_penalty`? 10.0

In [10]:
model_l1 = turicreate.linear_regression.create(training, target='price',
                                               features = all_features,
                                               l1_penalty = 10, l2_penalty = 0,
                                               validation_set = None, verbose = False)

model_l1.coefficients['value'].nnz()

18

***QUIZ QUESTION***
Also, using this value of L1 penalty, how many nonzero weights do you have? 18

# Limit the number of nonzero weights

What if we absolutely wanted to limit ourselves to, say, 7 features? This may be important if we want to derive "a rule of thumb" --- an interpretable model that has only a few features in them.

In this section, you are going to implement a simple, two phase procedure to achive this goal:
1. Explore a large range of `l1_penalty` values to find a narrow region of `l1_penalty` values where models are likely to have the desired number of non-zero weights.
2. Further explore the narrow region you found to find a good value for `l1_penalty` that achieves the desired sparsity.  Here, we will again use a validation set to choose the best value for `l1_penalty`.

In [11]:
max_nonzeros = 7

## Exploring the larger range of values to find a narrow range with the desired sparsity

Let's define a wide range of possible `l1_penalty_values`:

In [12]:
l1_penalty_values = np.logspace(8, 10, num=20)

Now, implement a loop that search through this space of possible `l1_penalty` values:

* For `l1_penalty` in `np.logspace(8, 10, num=20)`:
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list. When you call `linear_regression.create()` make sure you set `validation_set = None`
    * Extract the weights of the model and count the number of nonzeros. Save the number of nonzeros to a list.
        * *Hint: `model.coefficients['value']` gives you an SArray with the parameters you learned.  If you call the method `.nnz()` on it, you will find the number of non-zero parameters!* 

In [13]:
np_logspace = np.logspace(8, 10, num=20)

num_penality = len(np_logspace)

RSS = np.empty(num_penality)

for i in range(num_penality):
    model = turicreate.linear_regression.create(training, target = 'price',
                                                features = all_features,
                                                l1_penalty = np_logspace[i], l2_penalty = 0,
                                                validation_set = None, verbose = False)
    
    prediction = model.predict(validation)
    
    errors = prediction - validation['price']
    
    RSS[i] = np.dot(errors,errors)
    
    nnz = model.coefficients['value'].nnz()
    
    print ("For index " + str(i) + " number of non zeros: "+ str(nnz))
    
print (RSS)

For index 0 number of non zeros: 18
For index 1 number of non zeros: 18
For index 2 number of non zeros: 18
For index 3 number of non zeros: 18
For index 4 number of non zeros: 17
For index 5 number of non zeros: 17
For index 6 number of non zeros: 17
For index 7 number of non zeros: 17
For index 8 number of non zeros: 17
For index 9 number of non zeros: 16
For index 10 number of non zeros: 15
For index 11 number of non zeros: 15
For index 12 number of non zeros: 13
For index 13 number of non zeros: 12
For index 14 number of non zeros: 10
For index 15 number of non zeros: 6
For index 16 number of non zeros: 5
For index 17 number of non zeros: 3
For index 18 number of non zeros: 1
For index 19 number of non zeros: 1
[6.27492660e+14 6.28210517e+14 6.29176690e+14 6.30650083e+14
 6.32940229e+14 6.36268140e+14 6.41261198e+14 6.48983455e+14
 6.60962218e+14 6.77261521e+14 7.01046816e+14 7.37850623e+14
 7.96163110e+14 8.69018173e+14 9.66925692e+14 1.08186759e+15
 1.24492736e+15 1.38416149e+15 

Out of this large range, we want to find the two ends of our desired narrow range of `l1_penalty`.  At one end, we will have `l1_penalty` values that have too few non-zeros, and at the other end, we will have an `l1_penalty` that has too many non-zeros.  

More formally, find:
* The largest `l1_penalty` that has more non-zeros than `max_nonzeros` (if we pick a penalty smaller than this value, we will definitely have too many non-zero weights)
    * Store this value in the variable `l1_penalty_min` (we will use it later)
* The smallest `l1_penalty` that has fewer non-zeros than `max_nonzeros` (if we pick a penalty larger than this value, we will definitely have too few non-zero weights)
    * Store this value in the variable `l1_penalty_max` (we will use it later)


*Hint: there are many ways to do this, e.g.:*
* Programmatically within the loop above
* Creating a list with the number of non-zeros for each value of `l1_penalty` and inspecting it to find the appropriate boundaries.

In [14]:
l1_penalty_min = np_logspace[14]
l1_penalty_max = np_logspace[15]

print("l1_penalty_min: " + str(l1_penalty_min) + "\nl1_penalty_max: " + str(l1_penalty_max))

l1_penalty_min: 2976351441.6313133
l1_penalty_max: 3792690190.7322536


***QUIZ QUESTION.*** What values did you find for `l1_penalty_min` and `l1_penalty_max`, respectively? 

2976351441.6313133, 3792690190.7322536

## Exploring the narrow range of values to find the solution with the right number of non-zeros that has lowest RSS on the validation set 

We will now explore the narrow region of `l1_penalty` values we found:

In [15]:
l1_penalty_values = np.linspace(l1_penalty_min,l1_penalty_max,20)

* For `l1_penalty` in `np.linspace(l1_penalty_min,l1_penalty_max,20)`:
    * Fit a regression model with a given `l1_penalty` on TRAIN data. Specify `l1_penalty=l1_penalty` and `l2_penalty=0.` in the parameter list. When you call `linear_regression.create()` make sure you set `validation_set = None`
    * Measure the RSS of the learned model on the VALIDATION set

Find the model that the lowest RSS on the VALIDATION set and has sparsity *equal* to `max_nonzeros`.

In [16]:
len_l1 = len(l1_penalty_values)

RSS = np.empty(len_l1)

for i in range(len_l1):
    
    model = turicreate.linear_regression.create(training, target = 'price',
                                                features = all_features, 
                                                l1_penalty = l1_penalty_values[i], 
                                                l2_penalty = 0, validation_set = None,
                                                verbose = False)
    
    prediction = model.predict(validation)
    
    errors = prediction - validation['price']
    
    RSS[i] = np.dot(errors,errors)
    
    nnz = model.coefficients['value'].nnz()
    
    print ("For index " + str(i) + " number of non zeros: "+ str(nnz))
    
print (RSS)

For index 0 number of non zeros: 10
For index 1 number of non zeros: 10
For index 2 number of non zeros: 10
For index 3 number of non zeros: 10
For index 4 number of non zeros: 10
For index 5 number of non zeros: 10
For index 6 number of non zeros: 10
For index 7 number of non zeros: 10
For index 8 number of non zeros: 8
For index 9 number of non zeros: 8
For index 10 number of non zeros: 8
For index 11 number of non zeros: 7
For index 12 number of non zeros: 7
For index 13 number of non zeros: 7
For index 14 number of non zeros: 7
For index 15 number of non zeros: 6
For index 16 number of non zeros: 6
For index 17 number of non zeros: 6
For index 18 number of non zeros: 6
For index 19 number of non zeros: 6
[9.66925692e+14 9.74019450e+14 9.81188368e+14 9.89328342e+14
 9.98783211e+14 1.00847717e+15 1.01829878e+15 1.02824799e+15
 1.03461691e+15 1.03855474e+15 1.04323724e+15 1.04693749e+15
 1.05114763e+15 1.05599274e+15 1.06079953e+15 1.06570769e+15
 1.06946434e+15 1.07350455e+15 1.07763

***QUIZ QUESTIONS***
1. What value of `l1_penalty` in our narrow range has the lowest RSS on the VALIDATION set and has sparsity *equal* to `max_nonzeros`?
2. What features in this model have non-zero coefficients?

In [17]:
rss_min = RSS.min()

for i in range(num_penality):
    
    if RSS[i]==rss_min:
        
        print ("Penalty: " + str(np_logspace[i]) + "\nRSS: " + str(RSS[i]))

Penalty: 100000000.0
RSS: 966925692362085.6


In [18]:
rss_min = RSS[11:15].min()
print (rss_min) # min is RSS[11]

1046937488751711.0


In [19]:
print (l1_penalty_values[11])

3448968612.163437


In [20]:
model.coefficients.print_rows(num_rows=rows)

+------------------+-------+--------------------+--------+
|       name       | index |       value        | stderr |
+------------------+-------+--------------------+--------+
|   (intercept)    |  None | 240309.75931983022 |  None  |
|     bedrooms     |  None |        0.0         |  None  |
| bedrooms_square  |  None |        0.0         |  None  |
|    bathrooms     |  None | 13840.639957731408 |  None  |
|   sqft_living    |  None | 30.558358829776964 |  None  |
| sqft_living_sqrt |  None | 592.1994692127664  |  None  |
|     sqft_lot     |  None |        0.0         |  None  |
|  sqft_lot_sqrt   |  None |        0.0         |  None  |
|      floors      |  None |        0.0         |  None  |
|  floors_square   |  None |        0.0         |  None  |
|    waterfront    |  None |        0.0         |  None  |
|       view       |  None |        0.0         |  None  |
|    condition     |  None |        0.0         |  None  |
|      grade       |  None | 2265.1205255597065 |  None 